sim/priest/talents.go

```diff
@@ -60,9 +61,644 @@ func (priest *Priest) ApplyTalents() {
+			}
+		},
+	})
+
+	priest.RegisterAura(core.Aura{
+		Label:    "Mind Melt",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if result.Landed() && priest.MindSpike == spell {
+				priest.MindMeltProcAura.Activate(sim)
+				priest.MindMeltProcAura.AddStack(sim)
+			}
+		},
+	})
+}
+
+func (priest *Priest) applySinAndPunishment() {
+	if priest.Talents.SinAndPunishment == 0 {
+		return
+	}
+
+	priest.RegisterAura(core.Aura{
+		Label:    "SinAndPunishment",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			// check for critical mind flay tick
+			if result.Outcome.Matches(core.OutcomeCrit) && priest.MindFlayAPL.ClassSpellMask == int64(PriestSpellMindFlay) {
+
+				// reduce cooldown
+				remaining := max(0, time.Duration(*priest.Shadowfiend.CD.Timer)-time.Second*5)
+				priest.Shadowfiend.CD.Set(remaining)
+			}
+		},
+	})
+}
+
+func (priest *Priest) applyShadowyApparition() {
+	if priest.Talents.ShadowyApparition == 0 {
+		return
+	}
+
+	const spellScaling = 0.515
+	const levelScaling = 0.514
+
+	priest.ShadowyApparition = priest.RegisterSpell(core.SpellConfig{
+		ActionID:                 core.ActionID{SpellID: 87532},
+		MissileSpeed:             3.5,
+		ProcMask:                 core.ProcMaskEmpty, // summoned guardian, should not be able to proc stuff - verify
+		ClassSpellMask:           int64(PriestSpellShadowyApparation),
+		DamageMultiplier:         1,
+		DamageMultiplierAdditive: 1,
+		CritMultiplier:           priest.DefaultSpellCritMultiplier(),
+		Cast: core.CastConfig{
+			DefaultCast: core.Cast{
+				Cost: 0,
+				GCD:  0,
+			},
+		},
+
+		SpellSchool:      core.SpellSchoolShadow,
+		ThreatMultiplier: 1,
+
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			baseDamage := priest.ScalingBaseDamage*levelScaling + spellScaling*spell.SpellPower()
+
+			// snapshot values on spawn
+			dmgMulti := spell.DamageMultiplier
+			dmgMultiAdd := spell.DamageMultiplierAdditive
+
+			spell.WaitTravelTime(sim, func(sim *core.Simulation) {
+
+				oldMulti := spell.DamageMultiplier
+				oldAdd := spell.DamageMultiplierAdditive
+
+				// calculate dmg on hit, as the apparations profit from the debuffs on the target
+				// when they reach them
+				// spell and other modifiers are snapshotted when the apparations spawn
+				spell.DamageMultiplier = dmgMulti
+				spell.DamageMultiplierAdditive = dmgMultiAdd
+
+				result := spell.CalcDamage(sim, target, baseDamage, spell.OutcomeMagicHitAndCrit)
+				spell.DealDamage(sim, result)
+
+				// restore mods
+				spell.DamageMultiplier = oldMulti
+				spell.DamageMultiplierAdditive = oldAdd
```

Comment:
Usually all Attacker(Caster) mods are applied on CastComplete (when the apparition is cast) and target mods are applied when the damage is done (when the apparition hits the target). You dont need to snapshot any modifiers manually instead you do:
```
result := spell.CalcDamage(sim, target, baseDamage, outcom...)
spell.WaitTravelTime(sim, func(sim *core.Simulation) {
    spell.DealDamage(sim, result)
}
```
what this does is calculate the damage (+caster modifiers on the cast complete) and then after travel time deal the damage(applying target modifiers in the dealing part)

Reply:
Basically Caster mods are applied in the Calc part and target mods are applied in the Deal part